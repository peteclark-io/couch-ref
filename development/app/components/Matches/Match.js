'use strict'

import React from 'react';
import {Link} from 'react-router';
import {ThreeBounce} from 'better-react-spinkit';
import bootstrap from 'bootstrap/dist/css/bootstrap.css';
import classNames from 'classnames';

import styles from './styles.css';
import QuestionList from '../Questions/QuestionList';
import MatchRating from './MatchRating';
import RefereeRating from './RefereeRating';

const Match = React.createClass({

   propTypes: {
     match: React.PropTypes.shape({
         'id': React.PropTypes.string,
         'home': React.PropTypes.string,
         'away': React.PropTypes.string,
         'goalsHome': React.PropTypes.number,
         'goalsAway': React.PropTypes.number,
          'questions': React.PropTypes.arrayOf(
            React.PropTypes.shape({
              id: React.PropTypes.string,
              gid: React.PropTypes.string,
              group: React.PropTypes.string,
              asked: React.PropTypes.string,
              time: React.PropTypes.string,
              'questions': React.PropTypes.arrayOf(
                React.PropTypes.shape({
                  id: React.PropTypes.string
                }))
            }))
       })
   },

   render: function() {
      if (!this.props.match){
        return (
          <div className={styles.loading}>
            <ThreeBounce />
          </div>
        );
      }

      return (
         <div>
            <div className={bootstrap.row}>
               <div className={classNames(bootstrap['col-xs-12'], bootstrap['col-sm-8'])}>
                  <h2 className={styles['match-header']}>{this.props.match.home} {this.props.match.goalsHome}  -  {this.props.match.goalsAway} {this.props.match.away}</h2>
               </div>
               {
                 this.props.match.homeLineup === 0 ?
                 <div className={classNames(bootstrap['col-xs-12'], bootstrap['col-sm-4'])}>
                    <Link className={styles.link} to={`/teams/${this.props.match.id}`}><h2 className={styles['teams-link']}><small>Teams</small></h2></Link>
                 </div>
                 : null
               }
            </div>
            <MatchRating   id={this.props.match.id} />
            <RefereeRating id={this.props.match.id} />
            <QuestionList questions={this.props.match.questions} />
         </div>
      );
   }
});

export default Match;
