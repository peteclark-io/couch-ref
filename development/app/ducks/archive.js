'use strict';

import firebase from 'firebase';

import references from '../core/references';
import {createMatch, createQuestion, createStatistic} from '../core/mappers';

const LOAD_MATCH = 'couch-ref/archive/LOAD_MATCH'
const LOAD_QUESTION = 'couch-ref/archive/LOAD_QUESTION'
const LOAD_STATISTICS = 'couch-ref/archive/LOAD_STATISTICS'

export default function reducer(state = {}, action){
   switch(action.type){
      case LOAD_MATCH:
      return Object.assign({}, state, {
         matches: Object.assign({}, state.matches, {
            [action.match.id]: action.match
         })
      });

      case LOAD_QUESTION:
      return Object.assign({}, state, {
         questions: Object.assign({}, state.questions, {
            [action.question.id]: action.question
         })
      });

      case LOAD_STATISTICS:
      return Object.assign({}, state, {
         statistics: Object.assign({}, state.statistics, {
            [action.statistic.id]: action.statistic
         })
      });

      default:
      return state;
   }
};

export function loadMatch(match, redirect) {
   return (dispatch) => {
      var db = firebase.database();
      db.ref(references.archiveMatches + '/' + match).once('value').then((snap) => {
         if (!snap.exists()){
            redirect ? redirect() : null;
            return;
         }

         var archived = snap.val();

         dispatch({type: LOAD_MATCH, match: createMatch(archived)});

         if (!archived.questions){
            return;
         }

         archived.questions.map(q => {
            db.ref(references.archiveQuestions + '/' + q.id).once('value').then(snap => {
               var archivedQ = snap.val();
               dispatch({type: LOAD_QUESTION, question: createQuestion(archivedQ)});
            });
         });
      });
   };
}

export function loadQuestion(question, redirect) {
   return (dispatch) => {
      var db = firebase.database();
      db.ref(references.archiveQuestions + '/' + question).once('value').then((snap) => {
         if (!snap.exists()){
            redirect ? redirect() : null;
            return;
         }

         var archived = snap.val();
         dispatch({type: LOAD_QUESTION, question: createQuestion(archived)});
      });
   };
}

export function loadQuestionStatistics(question, redirect) {
   return (dispatch) => {
      var db = firebase.database();
      db.ref(references.archiveStatistics + '/' + question).once('value').then((snap) => {
         if (!snap.exists()){
            redirect ? redirect() : null;
            return;
         }

         var archived = snap.val();
         dispatch({type: LOAD_STATISTICS, statistic: createStatistic(archived)});

         db.ref(references.archiveQuestions + '/' + archived.id).once('value').then(snap => {
            var archivedQ = snap.val();
            loadMatch(archivedQ.match)(dispatch);
         });
      });
   };
}
