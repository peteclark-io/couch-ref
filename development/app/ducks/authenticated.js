'use strict';

const USER_AUTHENTICATED = 'couch-ref/authenticated/USER_AUTHENTICATED';

export default function reducer(state = false, action){
  switch(action.type){
    case USER_AUTHENTICATED:
      return action.authenticated; // should this just be true?
    default:
      return state;
  }
};

export function authenticated() {
  return {type: USER_AUTHENTICATED, authenticated: true};
};
