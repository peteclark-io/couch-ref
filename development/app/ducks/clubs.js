'use strict';

import _ from 'lodash';

const ADD_CLUBS = 'couch-ref/clubs/ADD_CLUBS'

export default function reducer(state = [], action){
   switch(action.type){
      case ADD_CLUBS:
      return _.sortBy(action.clubs, ['shortName']);

      default:
      return state;
   }
};

export function addClubs(clubs) {
   return {type: ADD_CLUBS, clubs: clubs};
}
