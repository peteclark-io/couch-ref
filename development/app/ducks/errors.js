'use strict';

const ADD_ERROR = 'couch-ref/errors/ADD_ERROR'

export default function reducer(state = {}, action){
  switch(action.type){
    case ADD_ERROR:
      return Object.assign({}, state, {
         [action.code]: {
            code: action.code,
            message: action.msg
         }
      });

    default:
      return state;
  }
};

export function addError(msg, code) {
  return {type: ADD_ERROR, msg: msg, code: code};
}
