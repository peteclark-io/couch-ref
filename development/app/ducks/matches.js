'use strict'

import _ from 'lodash';

const MATCH_UPDATE = 'couch-ref/matches/MATCH_UPDATE'
const ADD_MATCH = 'couch-ref/matches/ADD_MATCH'

function sort(arr){
  return _.sortBy(arr, ['fullTime', 'kickOff']);
}

export default function reducer(state = [], action){
  var match = action.match;

  switch(action.type){
    case MATCH_UPDATE:
      console.log('Action found!', match);
      return sort([
        ..._.differenceWith(state, [match], (val, compare) => {return val.id === compare.id}),
        match
      ]);

    case ADD_MATCH:
      console.log('Action found!', match);
      var update = state.filter((entry) => { return entry.id === match.id; }).length > 0;
      console.log('Is update', update);
      if (update){
         return sort([
           ..._.differenceWith(state, [match], (val, compare) => {return val.id === compare.id}),
           match
         ]);
      }

      return sort([...state, match]);
    default:
      return state;
  }
};

export function updateMatch(match) {
  return {type: MATCH_UPDATE, match: match};
}

export function addMatch(match) {
  return {type: ADD_MATCH, match: match};
}
