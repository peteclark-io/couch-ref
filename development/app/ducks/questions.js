'use strict';

const QUESTION_UPDATE = 'couch-ref/questions/MATCH_UPDATE'
const ADD_QUESTION = 'couch-ref/questions/ADD_MATCH'

export default function reducer(state = {}, action){
  switch(action.type){
    case QUESTION_UPDATE:
      return Object.assign({}, state, {
        [action.question.id]: action.question
      });

    case ADD_QUESTION:
      return Object.assign({}, state, {
        [action.question.id]: action.question
      });

    default:
      return state;
  }
};

export function updateQuestion(question) {
  return {type: QUESTION_UPDATE, question: question};
}

export function addQuestion(question) {
  return {type: ADD_QUESTION, question: question};
}
