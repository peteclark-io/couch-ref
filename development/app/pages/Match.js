'use strict'

import React from 'react';
import classNames from 'classnames';
import bootstrap from 'bootstrap/dist/css/bootstrap.css';

import styles from './MatchList.css';
import LiveScores from '../components/Matches/LiveScores';
import CouchRefHeader from '../components/Common/CouchRefHeader';

const Match = React.createClass({

  render: function() {
    return (
      <div className={bootstrap.container}>
        <div className={bootstrap.row}>
          <div className={classNames(bootstrap['col-xs-12'])}>
            <h2>Match</h2>
          </div>
        </div>
        <div className={styles.spacer}></div>
        <div className={bootstrap.row}>
          <div className={bootstrap['col-xs-12'], bootstrap['col-md-6'], bootstrap['col-lg-4']}>
            <LiveScores />
          </div>
        </div>
      </div>
    );
  }
});

export default Match;
