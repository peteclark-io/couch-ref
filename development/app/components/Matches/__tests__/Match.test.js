'use strict';

import React from 'react';
import Match from '../Match';
import renderer from 'react-test-renderer';
import {shallow, mount} from 'enzyme';

import reducer from '../../../ducks/reducers';
import {addMatch, updateMatch} from '../../../ducks/matches';
import {createStore} from 'redux';
import {Provider} from 'react-redux';

it('Displays loader when there is no data.', () => {
   const rendered = mount(
      <Match />
   );

   expect(rendered.find('.loading')).toHaveLength(1);
});

it('Displays match score, with no questions.', () => {
   var match = {id: '1001', home: 'Arsenal', away: 'West Ham', goalsHome: 1, goalsAway: 0, questions: []};
   const rendered = renderer.create(
      <Match match={match} />
   );

   expect(rendered.toJSON()).toMatchSnapshot();
});
