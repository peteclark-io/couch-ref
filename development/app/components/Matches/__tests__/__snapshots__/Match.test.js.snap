exports[`test Displays match score, with no questions. 1`] = `
<div>
  <div
    className="match-header row">
    <div
      className="col-xs-12 col-sm-8">
      <h2>
        Arsenal
         
        1
          -  
        0
         
        West Ham
      </h2>
    </div>
  </div>
  <div
    className="questions-list">
    <h1
      className="header">
      Live Questions
    </h1>
    <div
      className="row">
      <div
        className="col-xs-12">
        <h4
          className="spacer">
          No questions have been asked yet
        </h4>
      </div>
    </div>
  </div>
</div>
`;
